cirq/_compat_test.py

```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
+    """To ensure that module_a gets initialized only once."""
+    from cirq.testing._compat_test_data.module_a import module_b
+
+    assert module_b
+    from cirq.testing._compat_test_data.fake_a import module_b
+
+    assert module_b
+
+
+def _from_deprecated_import_sub_of_sub():
+    """Ensures that the deprecation warning level is correct."""
+    from cirq.testing._compat_test_data.module_a.module_b import module_c
+
+    assert module_c
+    from cirq.testing._compat_test_data.fake_a.module_b import module_c
+
+    assert module_c
+
+
+def _import_multiple_deprecated():
+    """Ensures that multiple deprecations play well together."""
+    from cirq.testing._compat_test_data.module_a.module_b import module_c
+
+    assert module_c
+    from cirq.testing._compat_test_data.fake_a.module_b import module_c
+
+    assert module_c
+    from cirq.testing._compat_test_data.fake_b import module_c
+
+    assert module_c
+
+
+def _new_module_in_different_parent():
+    from cirq.testing._compat_test_data.fake_ops import raw_types
+
+    assert raw_types
+
+
+def _find_spec_deprecated_multiple_times():
+    """to ensure the idempotency of the aliasing loader change"""
+    # sets up the DeprecationFinders
+    import importlib.util
+
+    # first import, the loader is the regular loader on the spec
+    assert importlib.util.find_spec("cirq.testing._compat_test_data.fake_a")
+    # second import it might be the aliasing loader already
+    assert importlib.util.find_spec("cirq.testing._compat_test_data.fake_a")
+
+
+def _import_parent_use_constant_from_deprecated():
+    """to ensure that module initializations set attributes correctly"""
+    # sets up the DeprecationFinders
+    import cirq.testing._compat_test_data
+
+    # the parent should have fake_a set on it as an attribute - just like
+    # a regular module import (e.g. cirq.ops)
+    # should have a DUPE_CONSTANT as its imported from the dupe submodule
+    assert cirq.testing._compat_test_data.fake_a.DUPE_CONSTANT == False
+
+
+def _import_deprecated_sub_use_constant():
+    """to ensure that submodule initializations set attributes correctly"""
+    # sets up the DeprecationFinders
+    import cirq.testing._compat_test_data.fake_a.dupe
+
+    # should have a DUPE_CONSTANT as its defined on it, set to False
+    assert cirq.testing._compat_test_data.fake_a.dupe.DUPE_CONSTANT == False
+
+
+def _import_deprecated_same_name_in_earlier_subtree():
+    from cirq.testing._compat_test_data.fake_a.sub.subsub.dupe import DUPE_CONSTANT
+
+    assert DUPE_CONSTANT
+
+
+def _import_top_level_deprecated():
+    from cirq.testing._compat_test_data.fake_freezegun import api
+
+    assert api
+
+
+# this is where the deprecation error should show where the deprecated usage
+# has occured, which is this file
+_deprecation_origin = ['_compat_test.py:']
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_a_deprecation_msg = [
+    'fake_a was used but is deprecated',
```

Comment:
Should this be?
```suggestion
    'cirq.testing._compat_test_data.fake_a was used but is deprecated',
```

Reply:
It could be, it's just a partial matching, so it made it less verbose

Reply:
Sure, but in this case I think it's an important part of the test.

Reply:
👍 added